### ISSUE TITLE
auto reconnect?

### ISSUE THREAD

**Issue by mstormi** [owntracks/android#255]
Hi,
after getting disconnected from the mqtt server due to network problems/radio coverage or whatever reason unrelated to owntracks, I noticed that owntracks remains disconnected on a number of occasions.
I believe having read somewhere that it's supposed to try reconnecting immediately, but that function either seems to be broken, or the algorithm is not well-suited to cover certain types of network outages.
If there's a way of generating a debug output to help you with, please let me know how this can be done. 
Using 0.5.20 beta on Android with a private cloud (own mqtt server) and no TLS. I've seen this happen with the legacy owntracks version on another Android device, too.

thanks in advance,
Markus

**Comment by binarybucks** [comment]
Hi, 

auto reconnect will not work on the legacy version for sure. 
It should be working on 0.5.20 but I too have my doubts since some days since I haven't seen a reconnect in case an error on the broker side occured. 

If you are able to get ADB logs from the app that include the specific behavior, I'd be more than grateful.

**Comment by binarybucks** [comment]
Also, could you elaborate under which conditions that occurred and if there were conditions where it connected automatically?

**Comment by mstormi** [comment]
Uh, hard to tell. Confirmed occurrences were when there was an IP 
outage, i.e. my mobile lost radio coverage for a couple of minutes.
Server was fine meanwhile.
I don't now for sure, though, if  my mobile's IP address remained the 
same when radio was available again.
As you might know, this can happen due to the way mobiles handle IP 
connectivity.
Maybe that's a case worth looking at: what's happening inside owntracks 
when the phone's IP address changes.
Maybe you could run owntracks off ADB and provoke that case ?
I'm sorry I don't know how to to that from ADB, and AFAIK owntracks 
cannot be made produce a simple debug logfile, can it ?
That would be a feature to help in many cases, I guess.
I believe (but that's unconfirmed) that non-reconnects happened, too, 
when there was a MQTT server restart, but I might be mistaken
that this had happened on a device that ran legacy owntracks (which as 
you say is expected behavior there).
I believe (but that's also unconfirmed) it also happened once when my 
phone ran on WLAN and there was another IP outage due to my phone
switching off WLAN during nighttime.

regards
Markus
Am 16.09.2015 um 19:22 schrieb Alexander Rust:

**Comment by mstormi** [comment]
sorry, forgot: I have not observed any occurrences when it re-connected 
automatically.
But of course that does not mean it didn't happen. Impossible to tell 
without a log.

Am 16.09.2015 um 19:22 schrieb Alexander Rust:

**Comment by jpmens** [comment]
If you're keeping track of what reaches the broker, you can actually tell and record disconnects server-side; the broker publishes an LWT when the client goes away. This JSON payload is published to the user's topic (e.g. `owntracks/user/device` and contains the timestamp (`tst`) of the original connection.

**Comment by binarybucks** [comment]
I'll see if I can add logging to a file. It would really help in a lot of
situations.

Could you please check the steps I posted earlier and report how the app
behaves during that provoked disconnect?

On 16 Sep 2015, at 20:08, Markus Storm <notifications@github.com> wrote:

Uh, hard to tell. Confirmed occurrences were when there was an IP
outage, i.e. my mobile lost radio coverage for a couple of minutes.
Server was fine meanwhile.
I don't now for sure, though, if my mobile's IP address remained the
same when radio was available again.
As you might know, this can happen due to the way mobiles handle IP
connectivity.
Maybe that's a case worth looking at: what's happening inside owntracks
when the phone's IP address changes.
Maybe you could run owntracks off ADB and provoke that case ?
I'm sorry I don't know how to to that from ADB, and AFAIK owntracks
cannot be made produce a simple debug logfile, can it ?
That would be a feature to help in many cases, I guess.
I believe (but that's unconfirmed) that non-reconnects happened, too,
when there was a MQTT server restart, but I might be mistaken
that this had happened on a device that ran legacy owntracks (which as
you say is expected behavior there).
I believe (but that's also unconfirmed) it also happened once when my
phone ran on WLAN and there was another IP outage due to my phone
switching off WLAN during nighttime.

regards
Markus
Am 16.09.2015 um 19:22 schrieb Alexander Rust:
>
> Also, could you elaborate under which conditions that occurred and if
> there were conditions where it connected automatically?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/owntracks/android/issues/255#issuecomment-140810942>.
>

—
Reply to this email directly or view it on GitHub
<https://github.com/owntracks/android/issues/255#issuecomment-140824518>.

**Comment by mstormi** [comment]
Pardon my ignorance, but what 'the steps I [i.e. you] posted earlier' do you refer to you mean to run it using ADB? Sorry, not being a developer myself, I don't know how to do that (but if you could give me a pointer, I'll try).

Another finding, I  just noticed that there are obviously many reconnects still going on. I haven't setup my broker (yet) to log LWTs as @jpmens  suggested,  but here's an excerpt from my MQTT server log that at least shows it's the client to disconnect, and there's many more entries like this one.. pretty much all the time that phone is in use.
Now the device below for sure is running legacy owntracks, running on WLAN during this time (which is not perfect, but for sure it's not THAT flakey ...) , and for sure all the reconnects were not initiated manually.

Now that you mentioned legacy owntracks isn't supposed to try re-connecting at all, that adds to the confusion ... . I also saw reconnect log entries for two more devices running the legacy Android version. Less reconnects with those, but no one-time events either (which I would expect to see like when you switch on the phone [or its WLAN link, respectively] in the morning).

Sep 17 07:39:22 pi mosquitto[22346]: New connection from 77.8.150.74 on port 8883.
Sep 17 07:39:23 pi mosquitto[22346]: Client smartphone-lutz disconnected.
Sep 17 07:39:23 pi mosquitto[22346]: New client connected from 77.8.150.74 as smartphone-lutz (c0, k900, u'lutz').
Sep 17 07:40:45 pi mosquitto[22346]: Socket error on client smartphone-lutz, disconnecting.
Sep 17 07:41:45 pi mosquitto[22346]: New connection from 77.8.150.74 on port 8883.
Sep 17 07:41:45 pi mosquitto[22346]: Client smartphone-lutz disconnected.
Sep 17 07:41:45 pi mosquitto[22346]: New client connected from 77.8.150.74 as smartphone-lutz (c0, k900, u'lutz').
Sep 17 07:42:11 pi mosquitto[22346]: Socket error on client smartphone-lutz, disconnecting.

**Comment by binarybucks** [comment]
Sorry, apparently my comment with the steps got lost. 

Please send logs with lwt messages and broker logs that that were taken for a device running OwnTracks 0.5.20. Mixing in OwnTracks legacy here does not help in debugging the problem. It is an old app with many bugs that have been fixed in the newer versions. Posting the broker logs that show the connects of OwnTracks legacy is very misleading as much of the connection code has been rewritten already. 

To clear things up a little, OwnTracks legacy was also supposed to reconnect automatically, but that function didn't work well. You might see some reconnect attempts but they were not reliable. 

Please do the following. 
Disconnect all devices running OwnTracks legacy or setup a new broker on a different port to test. 

Get ready to take the logs from your broker. 
Get ready to take lwt messages by subscribing to your broker with e.g. the mosquitto_sub utility to the topic owntracks/$user/$device. Supply the correct username and name of the device you will use. 
- Open the 0.5.20 app on one device and let it connect to the broker
- Open the notification drawer and locate the Owntracks notification
- Check that it says connected
- Enable flight mode
- The notification should now indicate that the app is disconnected
- Wait a minute or two
- Disable flight mode and check if the notification indicates any connection attempts

**Comment by mstormi** [comment]
Hmm. That test succeeded (or failed, if you like) because owntracks 
successfully re-connected, i.e. I didn't see the problem occur.
Tried a number of times. I did it with just-released 0.5.22. Thanks btw 
for releasing it. Not sure if you introduced any changes from .20 to 
this one (relevant w.r.t. re-connection behavior).

When testing,  I have not seen any LWT messages (or even any other MQTT 
message) neither at the time of disconnection nor at owntracks startup.
I even traced IP packets in parallel, too, so I'm pretty sure that there 
were none (and that I just didn't see them due to wrong subscription 
topic, bad mosquitto logging settings or whatever).
Shouldn't there have been a LWT msg at owntracks startup ? I'm not fully 
aware of how the mechanism is supposed to work (I'm new to MQTT, too - 
sorry),
but how does owntracks tell the broker about its last will, if not via 
LWT msg right after connecting to the broker ?.

So for now, I don't know what else to check out.
Did you/can you try my suggestion (to change the device's IP address 
while owntracks is running)?

regards
Markus

Am 17.09.2015 um 09:26 schrieb Alexander Rust:

**Comment by jpmens** [comment]
An LWT message is set up when the client connects to the broker. From this moment onwards, the broker checks availability of client (PING); as long as the client is alive nothing special happens (other than normal publish/subscribes). When the broker determines that the client has gone away (i.e. is dead), the *broker* then sends the LWT to subscribed clients *on behalf* of the dead client.

What you should do is to `mosquitto_sub -v -t 'owntracks/#'` against your broker and leave that running.

Then launch OwnTracks app, and hit PUBLISH now. You should see that message arrive in the terminal doing the `_sub`.

After a moment or two, disable WiFi or whatever by e.g. turning on Airplane Mode. That will, from the broker's point of view, make the client seem dead. After a minute or two, you should see that LWT in your `mosquitto_sub` terminal (we publis a JSON payload with `{"_type":"lwt", ...}`.

**Comment by mstormi** [comment]
Thanks for explaining. Yes, I had been listening for 'owntracks/#' already. 

My most recent test showed three things:
1) I didn't see *any* lwt messages. I don't know why. Do we know owntracks properly registers its last will ?  (that 'will' I don't see in any of the subscribed-to messages, do I ?)
Would I have need to wait more than 5 mins for the timeout to happen? I did see a couple of PINGREQ/PINGRESP.
Is my broker-side timeout set too large ? I didn't find where to change that. 
During another test, I did see lwt from another device. But for this device I'm testing with, seems I can't get the broker to send lwt at all.
2) unlike my suspicion, owntracks does properly handle the case when IP address changes. See log below: it reconnects, gets disconnected by broker, then successfully re-connects. So an IP address change wasn't the reason for my original problem.
3) something strange with the client name, see output below. I used a '-' between "smartphone" and "markus" in owntracks, but the broker does not show the '-' in its logs (but the published messages do include it). Might be a cosmetic problem of the output function or a real one. I changed to a ID with no '-' (or any other special character), but that did not make any noteable difference.

Sep 19 22:50:36 pi mosquitto[7523]: New connection from 80.187.108.40 on port 8883.
Sep 19 22:50:36 pi mosquitto[7523]: Client markus/smartphonemarkus already connected, closing old connection.
Sep 19 22:50:36 pi mosquitto[7523]: Client markus/smartphonemarkus disconnected.
Sep 19 22:50:36 pi mosquitto[7523]: New client connected from 80.187.108.40 as markus/smartphonemarkus (c1, k3600, u'markus').
Sep 19 22:50:37 pi mosquitto[7523]: New connection from 77.8.150.74 on port 8883.
Sep 19 22:50:37 pi mosquitto[7523]: Client markus/smartphonemarkus already connected, closing old connection.
Sep 19 22:50:37 pi mosquitto[7523]: Client markus/smartphonemarkus disconnected.
Sep 19 22:50:37 pi mosquitto[7523]: New client connected from 77.8.150.74 as markus/smartphonemarkus (c1, k3600, u'markus').

**Comment by jpmens** [comment]
It looks to me as though this is working well. Your app was connected from an address and moved into a different address. Upon reconnection the broker realized the client was already connected (under that clientid) and kicked the first. So far so good and reconnects appear to be working for you, which is good.

Regarding the `_` in the clientID; I really don't know; sounds a bit strange.

The LWT will be published by the broker only when the client dies. Have you tried the "airplane mode" trick?

**Comment by mstormi** [comment]
Am 20.09.2015 um 10:27 schrieb JP Mens:
>
> It looks to me as though this is working well. Your app was connected 
> from an address and moved into a different address. Upon reconnection 
> the broker realized the client was already connected (under that 
> clientid) and kicked the first. So far so good and reconnects appear 
> to be working for you, which is good.
>
Yes, during all my testing, reconnects always worked. I'm now entitled 
to believe that there's no problem in owntracks re-connection code
and what had hit me was just the artefact of an intermittent, 
non-reproduceable situation. Maybe, when it happened on the move and
I *thought* I had radio coverage again, I didn't really have it (or 
radio was there, but IP connectivity still was not).
So regarding the original issue, let's consider that done, and close it. 
Sorry for the workload I created!
>
> Regarding the |_| in the clientID; I really don't know; sounds a bit 
> strange.
>
FWIW, I renamed all my devices to not have '-' included. But as I now 
still don't see LWTs, this is apparently unrelated.
Probably just a cosmetic bug in mosquitto output.
>
> The LWT will be published by the broker only when the client dies. 
> Have you tried the "airplane mode" trick?
>
Now this one is still bugging me.
Yes of course, I went into airplane mode.  I also ran tcpdump and I'm 
sure there were no IP packets arriving from the device.
I believe there's no LWTs because the broker does not (yet) believe the 
client is dead, although I tried to wait
up to ten minutes in airplane mode. I do have seen LWTs for another 
device, so those do work in principle.
But I don't know how much time had passed from that device going offline 
until the broker sent LWT for it.
Do you know if this timeout can somehow be tweaked in mosquitto ? I 
haven't seen any such setting
The common TCP timeout on Linux should be 180 seconds, although I'm not 
sure if that was increased in Android.

**Comment by binarybucks** [comment]
the app connects with a keepalive value of 3600 seconds. If the broker
doesn't hear from the client during that period (plus 0.5x that I think) he
considers the client dead.
As a result, it might take more than your 10 minute to receive a LWT.

Mit freundlichen Grüßen / With kind regards
Alexander Rust

On 20 Sep 2015, at 11:28, Markus Storm <notifications@github.com> wrote:



Am 20.09.2015 um 10:27 schrieb JP Mens:
>
> It looks to me as though this is working well. Your app was connected
> from an address and moved into a different address. Upon reconnection
> the broker realized the client was already connected (under that
> clientid) and kicked the first. So far so good and reconnects appear
> to be working for you, which is good.
>
Yes, during all my testing, reconnects always worked. I'm now entitled
to believe that there's no problem in owntracks re-connection code
and what had hit me was just the artefact of an intermittent,
non-reproduceable situation. Maybe, when it happened on the move and
I *thought* I had radio coverage again, I didn't really have it (or
radio was there, but IP connectivity still was not).
So regarding the original issue, let's consider that done, and close it.
Sorry for the workload I created!
>
> Regarding the |_| in the clientID; I really don't know; sounds a bit
> strange.
>
FWIW, I renamed all my devices to not have '-' included. But as I now
still don't see LWTs, this is apparently unrelated.
Probably just a cosmetic bug in mosquitto output.
>
> The LWT will be published by the broker only when the client dies.
> Have you tried the "airplane mode" trick?
>
Now this one is still bugging me.
Yes of course, I went into airplane mode. I also ran tcpdump and I'm
sure there were no IP packets arriving from the device.
I believe there's no LWTs because the broker does not (yet) believe the
client is dead, although I tried to wait
up to ten minutes in airplane mode. I do have seen LWTs for another
device, so those do work in principle.
But I don't know how much time had passed from that device going offline
until the broker sent LWT for it.
Do you know if this timeout can somehow be tweaked in mosquitto ? I
haven't seen any such setting
The common TCP timeout on Linux should be 180 seconds, although I'm not
sure if that was increased in Android.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/owntracks/android/issues/255#issuecomment-141758166>.
>

—
Reply to this email directly or view it on GitHub
<https://github.com/owntracks/android/issues/255#issuecomment-141762697>.

**Comment by binarybucks** [comment]
also note, that the keepalive interval and dead client detecting has
nothing to do with TCP timeouts or keepalives. MQTT was designed with
mobile connections in mind so it does not need a persistent TCP connection
to maintain sessions.

Mit freundlichen Grüßen / With kind regards
Alexander Rust

On 20 Sep 2015, at 11:28, Markus Storm <notifications@github.com> wrote:



Am 20.09.2015 um 10:27 schrieb JP Mens:
>
> It looks to me as though this is working well. Your app was connected
> from an address and moved into a different address. Upon reconnection
> the broker realized the client was already connected (under that
> clientid) and kicked the first. So far so good and reconnects appear
> to be working for you, which is good.
>
Yes, during all my testing, reconnects always worked. I'm now entitled
to believe that there's no problem in owntracks re-connection code
and what had hit me was just the artefact of an intermittent,
non-reproduceable situation. Maybe, when it happened on the move and
I *thought* I had radio coverage again, I didn't really have it (or
radio was there, but IP connectivity still was not).
So regarding the original issue, let's consider that done, and close it.
Sorry for the workload I created!
>
> Regarding the |_| in the clientID; I really don't know; sounds a bit
> strange.
>
FWIW, I renamed all my devices to not have '-' included. But as I now
still don't see LWTs, this is apparently unrelated.
Probably just a cosmetic bug in mosquitto output.
>
> The LWT will be published by the broker only when the client dies.
> Have you tried the "airplane mode" trick?
>
Now this one is still bugging me.
Yes of course, I went into airplane mode. I also ran tcpdump and I'm
sure there were no IP packets arriving from the device.
I believe there's no LWTs because the broker does not (yet) believe the
client is dead, although I tried to wait
up to ten minutes in airplane mode. I do have seen LWTs for another
device, so those do work in principle.
But I don't know how much time had passed from that device going offline
until the broker sent LWT for it.
Do you know if this timeout can somehow be tweaked in mosquitto ? I
haven't seen any such setting
The common TCP timeout on Linux should be 180 seconds, although I'm not
sure if that was increased in Android.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/owntracks/android/issues/255#issuecomment-141758166>.
>

—
Reply to this email directly or view it on GitHub
<https://github.com/owntracks/android/issues/255#issuecomment-141762697>.

**Comment by ckrey** [comment]
The keepalive interval is set by the client. The server should send the LWT if it the interval plus 50% passed without receiving any MQTT packet from the client (e.g. after 90 sec if keepalive was 60 sec.

From the spec:
> The actual value of the Keep Alive is application specific; typically this is a few minutes. The maximum value is 18 hours 12 minutes and 15 seconds.

For example, `mosquitto_sub` uses a default of 60 sec.

In addition, I have to say that the mosquitto broker sends an LWT message if the TCP connection is closed/aborts without sending an MQTT-Disconnect before

**Comment by mstormi** [comment]
Ah, thanks for the explanation. That's what the setting in Connection - 
Parameters is for, right?
For a test, I've set it to 60 and yes, after 90 seconds, I finally got 
to see my LWT.
So that's behaving as expected, i.e. no further investigation required.
Thanks for your input and patience!

best regards
Markus

PS: maybe add these infos to the booklet ?

Am 20.09.2015 um 18:32 schrieb Alexander Rust:
> the app connects with a keepalive value of 3600 seconds. If the broker
> doesn't hear from the client during that period (plus 0.5x that I 
> think) he
> considers the client dead.
> As a result, it might take more than your 10 minute to receive a LWT.
>
> Mit freundlichen Grüßen / With kind regards
> Alexander Rust
>

**Comment by mstormi** [comment]
Hi,
I encountered another instance of "non-reconnection".
owntracks showed it couldn't connect because it was unable to resolve the DNS name of my server.
Now since I'm using a DNS name for my broker's hostname (using a DynDNS like service to have a dynamic resolution, a common setup that probably most owntracks users will use, too), failure to translate the DNS name into an IP address will obviously cause owntracks fail connecting without even
trying to connect once.
But in my situation, owntracks remained in that state for hours. My IP stayed the same all the time, DNS server "returned", yet owntracks apparently didn't re-try re-resolving the name and connecting.
Can you please check the code if you try re-connecting after DNS resolution failures ?
There might be negative DNS caching in effect: if the resolver returns NXDOMAIN (i.e. "no such host"), Android or eventually owntracks itself may store and keep re-using this "answer" longer than they actually should (causing a persistent failure to connect even when IP connectivity and/or the server have become available again).

regards
Markus

**Comment by jpmens** [comment]
If negative caching is happening, that is ocurring at the OS level -- not in our app.
Would you please run the following command and paste its output?

```
dig your-domain soa +multi
```

(If need be you can X out the domain part on the left.)

**Comment by watou** [comment]
I can report that OwnTracks fails to reconnect automatically when DynDns changes my broker's IP address, but seems to reconnect when I re-save my connection settings manually.

**Comment by mstormi** [comment]
;; AUTHORITY SECTION:
selfhost.de.            2560 IN SOA pri.selfhost.de. hostmaster.selfhost.de. (
                                1443634759 ; serial
                                16384      ; refresh (4 hours 33 minutes 4 seconds)
                                2048       ; retry (34 minutes 8 seconds)
                                1048576    ; expire (1 week 5 days 3 hours 16 minutes 16 seconds)
                                2560       ; minimum (42 minutes 40 seconds)
                                )


That was a query to the primary NS.
So Android (or anyone to use the resolver library) should do negative caching for no more than 2560 seconds, right? My owntracks remained unconnected for much longer than that.
FWIW, my dialup's A record has a TTL of 60 seconds.

**Comment by jpmens** [comment]
Thank  you for that data, and yes, you are correct: negative caching for max 2560s. Anyway, assuming the domain you use (e.g. `mymqttbroker.selfhost.de` is available this shouldn't be an issue.

We will investigate. I agree it sounds like a bug in the reconnection code, also because of what @watou says.

I am opening a new issue for this.

**Comment by binarybucks** [comment]
Could you test if this has been fixed in 0.5.23 beta?

**Comment by mstormi** [comment]
Well, I don't know how exactly to reproduce the issue. But I installed 
0.5.23 beta right away 4 days ago and will report if I encounter issues. 
None so far.

Am 11.10.2015 um 15:10 schrieb Alexander Rust:

**Comment by nbartels** [comment]
I run in this (or a similar) problem, too. The error message I see is:

> Private MQTT - Error: failed to connect to host/ip (port 8883) after 30000ms: connect failed: ENETUNREACH (Network is unreachable)

I think the interesting part is this happens if I don't use my smartphone for a while. But I have to go to the owntracks settings and confirm the connection to force wontracks to reconnect. Then everythings fine. 

I'm using 0.6.14 and mosquitto as backend.

**Comment by arnoo** [comment]
I have the same issue : Owntracks stays disconnected until I resave the connection settings. Using 0.6.14

**Comment by yuchangyuan** [comment]
I have similar issue on 0.6.14, the message show on `Connection` of `Preferences` is 

> Private MQTT - Error: failed to connect to /xxx.xxx.xxx.xxx(port yyyy) after 30000ms: isConnected failed: ENETUNREACH(Network is unreachable)

where xxx.xxx.xxx.xxx is my MQTT server IP and yyyy is the server port.
